### ISSUE TITLE
BerkeleyDB memory exists beyond rm -rf

### ISSUE THREAD

**Issue by okram** [thinkaurelius/titan#251]
I forgot to add unique(OUT) so I rm -rf'd the /tmp/test-index directory. I then try to reinstantiate a new graph using TitanFactory.open(). Note that the directory was removed but it still 'sees' the old database schema.

``` text
gremlin> g.addVertex([service:'computers',name:'marko'])
Not an out-unique key: service
Display stack trace? [yN]
// IN ANOTHER WINDOW I sudo rm -rf /tmp/test-index
gremlin> g = TitanFactory.open('/tmp/test-index')
==>titangraph[local:/tmp/test-index]
gremlin> g.makeType().name('service').indexed(Vertex.class).unique(OUT).dataType(String.class).makePropertyKey()
The given value is already used as a property
Display stack trace? [yN]
gremlin> quit
/usr/local/titan-all-0.3.0$ bin/gremlin.sh

         \,,,/
         (o o)
-----oOOo-(_)-oOOo-----
gremlin> g = TitanFactory.open('/tmp/test-index')
==>titangraph[local:/tmp/test-index]
gremlin> g.makeType().name('service').indexed(Vertex.class).unique(OUT).dataType(String.class).makePropertyKey()
==>v[36028797018963978]
gremlin>
```

**Comment by mbroecheler** [comment]
It seems like BDB caches the database state in the jvm somehow. Exiting the
gremlin shell and re-starting it solves the problem.

On Thu, May 2, 2013 at 8:49 AM, Marko A. Rodriguez <notifications@github.com

> wrote:
> 
> I forgot to add unique(OUT) so I rm -rf'd the /tmp/test-index directory. I
> then try to reinstantiate a new graph using TitanFactory.open(). Note that
> the directory was removed but it still 'sees' the old database schema.
> 
> gremlin> g.addVertex([service:'computers',name:'marko'])
> Not an out-unique key: service
> Display stack trace? [yN]
> // IN ANOTHER WINDOW I sudo rm -rf /tmp/test-index
> gremlin> g = TitanFactory.open('/tmp/test-index')
> ==>titangraph[local:/tmp/test-index]
> gremlin> g.makeType().name('service').indexed(Vertex.class).unique(OUT).dataType(String.class).makePropertyKey()
> The given value is already used as a property
> Display stack trace? [yN]
> gremlin> quit
> /usr/local/titan-all-0.3.0$ bin/gremlin.sh
> 
> ```
>      \,,,/
>      (o o)
> ```
> 
> -----oOOo-(_)-oOOo-----
> gremlin> g = TitanFactory.open('/tmp/test-index')
> ==>titangraph[local:/tmp/test-index]
> gremlin> g.makeType().name('service').indexed(Vertex.class).unique(OUT).dataType(String.class).makePropertyKey()
> ==>v[36028797018963978]
> gremlin>
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/thinkaurelius/titan/issues/251
> .

## 

Matthias Broecheler
http://www.matthiasb.com

**Comment by okram** [comment]
Yea. That is the solution, but would be "nice" if it was smart. Not crazy important, but for some people they will get confused.

**Comment by mbroecheler** [comment]
I tried to make BDB "aware" of the data being deleted, but with no luck. My guess is that its still memory-mapped somewhere and therefore effectively cached.
Anyhow, I think there is a better way for dealing with this use case. Titan has the functionality to clear all backend storage (including any configured indexes) with one call. It's been hidden thus far so I added a utility class to make it available. Use **TitanCleanup.clear(graph)** to clear out all data for graph. Make sure the graph has been shutdown before calling this method. Now, you don't have to jump terminal windows anymore. This has been pushed to titan03.

``` groovy
gremlin> g = TitanFactory.open('/tmp/titan');
==>titangraph[local:/tmp/titan]
gremlin> g.addVertex([name:'Marko'])
==>v[4]
gremlin> g.V
==>v[4]
gremlin> g.commit()
==>null
gremlin> g.V
==>v[4]
gremlin> g.shutdown()
==>null
gremlin> com.thinkaurelius.titan.core.util.TitanCleanup.clear(g)
==>null
gremlin> g = TitanFactory.open('/tmp/titan');
==>titangraph[local:/tmp/titan]
gremlin> g.V
gremlin> 
```
